apis: drop check for volumes with user namespaces

giuseppe opened the pull request:
The second phase of user namespaces support was related to supporting only stateless pods.  Since the changes were accepted for the KEP, now the scope is extended to support stateful pods as well.  Remove the check that blocks creating PODs with volumes when using user namespaces.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?


/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Now it is possible to use pods with volumes and user namespaces.  The feature gate was renamed from UserNamespacesStatelessPodsSupport to UserNamespacesSupport
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:


KEP: https://github.com/kubernetes/enhancements/issues/127

giuseppe committed:
apis: drop check for volumes with user namespaces

The second phase of user namespaces support was related to supporting
only stateless pods.  Since the changes were accepted for the KEP, now
the scope is extended to support stateful pods as well.  Remove the
check that blocks creating PODs with volumes when using user namespaces.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe commented:
@rata @mrunalp PTAL

rata commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -21780,8 +21780,8 @@ func TestValidateHostUsers(t *testing.T) {
-		name:    "hostUsers=false - unsupported volume",
-		success: false,
+		name:    "hostUsers=false - stateful volume",
+		success: true,
```
nit, do we want to update the comment below that says: `		// It should ignore unsupported volumes with hostUsers=true.` ? There are no unsupported volume types now

(line 21798)

giuseppe replied:
yes, thanks for noticing it.  I'll push a fixed version

rata reviewed (approved):
@giuseppe LGTM, thanks! Left one simple comment about updating a comment in the tests, but is super minor :)

giuseppe committed:
features: rename UserNamespacesStatelessPodsSupport

now it is called UserNamespacesSupport since all kind of volumes are
supported.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rata reviewed (approved):
LGTM, thanks!

rata commented:
@thockin friendly ping?

thockin commented:
Things not assigned to me fall into lots of cracks. :)

thockin commented:
I had to go back to the KEP, and I have a question on the idmapped approach.  I think one of the primary questions we need to ask is "what happens in the case of a container escape?"  Userns means that (probably) such an escape results in code execution as the host UID, so even if running as 0 in container, it's non-zero in host.  Yay.

What about volumes?  IIUC (please confirm) the host-side UID/GID on files will be in the 0-64K range, which is never used for userns, right?  So if I escape my container, I can't read those files any more, or anyone else's files (unless they allow "other" read).  Is that correct?

Is there anyothing we could do to limit even those "other" reads in case of an escape?

thockin commented on pkg/features/kube_features.go:
```diff
@@ -793,7 +793,7 @@ const (
 	// alpha: v1.25
 	//
 	// Enables user namespace support for stateless pods.
-	UserNamespacesStatelessPodsSupport featuregate.Feature = "UserNamespacesStatelessPodsSupport"
```
This is effectively a breaking change.  Can we make this add a new gate and then support both as equivalents?  Or only support stateless if stateless gate is set and stateless + stateful in new gate?

rata replied:
I'm not sure that is a good idea, the user can have some surprises if we do that. Allow me to elaborate.

If the `UserNamespacesStatelessPodsSupport` feature gate also enables stateful pods, then the user can _very_ easily be surprised about that and is not something intuitive. Furthermore, as the requirements for stateless pods also changed (kernel support, container runtime support also needs changes, etc.) it might even be the case that with the stack they are using userns is not supported for stateless pods either. So it is not, if the feature gate does that, then things will "just work" for users of stateless pods.

Doing only stateless if only the stateless feature gate is active is tricky and not sure the churn is worth it. Because of what I already mentioned (things will "just work" for users of stateless pods) and because we will need to make the validation conditional on a feature gate. That is tricky, as we have the same pod.spec field name for stateless and stateful (pod.spec.hostUsers), and we can't check for the feature gate in that part of the code: https://github.com/kubernetes/kubernetes/pull/111090#discussion_r933556896, we might need to modify the ForUpdate/similar functions, if that is even possible to do.

If there were other reasons, it can be worth exploring that path. But IMHO, the pain is not worth the effort. This is an alpha feature, that was reworked and has new requirements from the container runtime and the kernel now, let's take this chance to change the feature gate and make sure users that opt-in are aware of the new requirements introduced.

@thockin If you feel strongly about doing any of those things, we can of course do it. In the end we want this to progress, so if you feel strongly about it, we will of course do it. But wanted to highlight those things in case you weren't aware

thockin replied:
OK.  If we release-note it as "alpha feature gate UserNamespacesStatelessPodsSupport has been abandoned" and "alpha feature gate UserNamespaces is new", I guess I can live with that.

rata commented:
@thockin
  
> What about volumes? IIUC (please confirm) the host-side UID/GID on files will be in the 0-64K range, which is never used for userns, right? So if I escape my container, I can't read those files any more, or anyone else's files (unless they allow "other" read). Is that correct?

Correct :)

> Is there anyothing we could do to limit even those "other" reads in case of an escape?

Not with userns. We need other security layers to cover that, probably apparmor can help. I want to explore landlock too (it can have several advantages, and I want to see if we can enable some things transparently in OCI runtimes). But those are completely orthogonal things.

rata commented:
/assign thockin

thockin reviewed (commented):
Thanks!

/lgtm
/approve
